sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -440,8 +440,12 @@ trait RuntimeReplaceable extends Expression {
   // are semantically equal.
   override lazy val canonicalized: Expression = replacement.canonicalized
 
-  final override def eval(input: InternalRow = null): Any =
-    throw QueryExecutionErrors.cannotEvaluateExpressionError(this)
+  final override def eval(input: InternalRow = null): Any = {
+    // For convenience, we allow to evaluate `RuntimeReplaceable` expressions, in case we need to
+    // get a constant from foldable expression before the query execution starts.
+    assert(input == null)
```

Comment:
Sorry, could you clarify why do you require `null` instead of just bypass it to `replacement.eval()`

Reply:
If the premise here must be that `input == null`, then would it be better to throw a meaningful QueryExecutionError instead of an AssertionError when input is not null?

Reply:
Ideally people should never eagerly evaluate a `RuntimeReplaceable`, the only case I see is evaluating foldable expressions. I think it's better to only allow it, to make sure that there is no `RuntimeReplaceable` leaked to the runtime execution.